metadatastore/commands.py

```diff
@@ -39,167 +99,512 @@ def inner(*args, **kwargs):
+
+    Raises if no EventDescriptors exist.
+
+    Parameters
+    ----------
+    runstart : doc.Document or dict or str
+        The RunStart to get the RunStop for.  Can be either
+        a Document/dict with a 'uid' key or a uid string
+
+    Returns
+    -------
+    event_descriptors : list
+        A list of EventDescriptor documents
+
+    Raises
+    ------
+    NoEventDescriptors
+        If no EventDescriptor documents exist for the given RunStart
+    """
+    # normalize the input and get the runstart oid
+    runstart_uid = doc_or_uid_to_uid(runstart)
+    oid = _RUNSTART_UID_to_OID_MAP[runstart_uid]
+
+    # query the database for any event descriptors which
+    # refer to the given runstart
+    descriptors = EventDescriptor._get_collection().find(
+        {'run_start_id': oid})
+
+    # loop over the found documents, cache, and dereference
+    rets = [_cache_descriptor(descriptor) for descriptor in descriptors]
+
+    # if nothing found, raise
+    if not rets:
+        raise NoEventDescriptors("No EventDescriptors exists")
+
+    # return the list of event descriptors
+    return rets
+
+
+def fetch_events_generator(descriptor):
+    """A generator which yields all events from the event stream
+
+    Parameters
+    ----------
+    descriptor : doc.Document or dict or str
+        The RunStart to get the RunStop for.  Can be either
+        a Document/dict with a 'uid' key or a uid string
+
+    Yields
+    ------
+    event : doc.Document
+        All events for the given EventDescriptor from oldest to
+        newest
+    """
+    descriptor_uid = doc_or_uid_to_uid(descriptor)
+    descriptor = descriptor_given_uid(descriptor_uid)
+    col = Event._get_collection()
+
+    oid = _EVENTDESC_UID_to_OID_MAP[descriptor_uid]
+
+    ev_cur = col.find({'descriptor_id': oid},
+                      sort=[('time', 1)])
+
+    for ev in ev_cur:
+        # ditch the ObjectID
+        ev.pop('_id')
+        # pop the descriptor oid
+        ev.pop('descriptor_id')
+        # replace it with the defererenced descriptor
+        ev['descriptor'] = descriptor
+        # pop the data
+        data = ev.pop('data')
+        # replace it with the friendly paired dicts
+        ev['data'], ev['timestamps'] = [{k: v[j] for k, v in data.items()}
+                                        for j in range(2)]
+        # wrap it our fancy dict
+        ev = doc.Document('Event', ev)
+
+        yield ev
+
+
+def fetch_events_table(descriptor):
+    """All event data as tables
+
+    .. warning
+
+       This function has a local import of `pandas` and will only work if
+       `pandas` is installed.
+
+    Parameters
+    ----------
+    descriptor : dict or uid
+        The EeventDestriptor to get the events for.  Can be either
+        a dict or a uid.
+
+    Returns
+    -------
+    descriptor : dict
+        EventDescriptor document
+
+    data_table : DataFrame
+        All of the data as a DataFrame indexed on sequence number
```

Comment:
It is going to be challenging to combine events from multiple runs if these are indexed on sequence number because sequence numbers are degenerate across different run_start_uids.  Are sequence numbers also degenerate between the `event descriptors` of one RunStart?

As such, I would advocate for indexing these on event time, not sequence number.